solidity/contracts/BancorNetwork.sol

```diff
@@ -450,6 +405,191 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
         return claimAndConvertFor2(_path, _amount, _minReturn, msg.sender, _affiliateAccount, _affiliateFee);
     }
 
+    /**
+      * @dev validates msg.value and prepares the conversion source token for the conversion
+      * 
+      * @param _sourceToken source token of the first conversion step
+      * @param _smartToken  smart token of the first conversion step
+      * @param _amount      amount to convert from, in the source token
+    */
+    function handleSourceToken(IERC20Token _sourceToken, ISmartToken _smartToken, uint256 _amount) private {
+        bool isSourceETH = etherTokens[_sourceToken];
+
+        // validate msg.value
+        if (msg.value > 0)
+            require(msg.value == _amount && isSourceETH);
+        else
+            require(msg.value == 0);
+
+        IBancorConverter firstConverter = IBancorConverter(_smartToken.owner());
+        bool isNewerConverter = isV27OrHigherConverter(firstConverter);
+
+        // ETH / EtherToken
+        if (isSourceETH) {
+            // handle ETH
+            if (msg.value > 0) {

```

Comment:
I think that this logic is wrong.
Instead of if/else on `msg.value > 0`, we should do if/else on `isNewerConverter`.
It will subsequently change the entire implementation of the current if/else clause.
Also, what if `_sourceToken == address(0)`?

Reply:
I actually don't see an issue but let's go over it together, will be more efficient

Reply:
@barakman cleaned up